sklearn/tests/test_isotonic.py

```diff
@@ -485,6 +485,16 @@ def test_isotonic_dtype():
             assert res.dtype == expected_dtype
 
 
+def test_isotonic_mismatched_dtype():
+    reg = IsotonicRegression()
+    y = [2, 1, 4, 3, 5]
+    for dtype in (np.int32, np.int64, np.float64):
+        y_np = np.array(y, dtype=dtype)
+        X = np.arange(len(y)).astype(np.float32)
```

Comment:
Nit: Better to make `X`  2D instead of relying on `check_array` to convert it,

```suggestion
        X = np.arange(len(y), dtype=np.float32).reshape((-1, 1))
```

Reply:
I am not very sure about reshaping here because I wonder if the not-reshaped input is intended as part of the test. So I kept it as 1D like previous.

Reply:
Exactly, the current version make one wonder whether you want to also check that the input is reshaped to 2D. However that's not the intention of this test as far as I can tell? Therefore better to do one thing, and only check dtype.

Reply:
I see, I merged your suggestions into the test. Thanks.